google-cloud-pubsub/lib/google/cloud/pubsub/subscriber.rb

```diff
@@ -0,0 +1,217 @@
+# Copyright 2017 Google Inc. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+require "google/cloud/pubsub/service"
+require "google/cloud/errors"
+require "monitor"
+require "forwardable"
+require "concurrent"
+
+module Google
+  module Cloud
+    module Pubsub
+      ##
+      # # Subscriber
+      #
+      class Subscriber
+        include MonitorMixin
+
+        ##
+        # @private Implementation attributes.
+        attr_reader :request_queue, :output_enum, :thread_pool, :service
+
+        ##
+        # Subscriber attributes.
+        attr_reader :callback, :subscription, :deadline, :threads
+
+        ##
+        # @private Create an empty {Subscriber} object.
+        def initialize callback, subscription, deadline, threads, service
+          @request_queue = nil
+          @output_enum = nil
+          @callback = callback
+          @subscription = subscription
+          @deadline = deadline
+          @threads = threads || [2, Concurrent.processor_count * 2].max
+          @service = service
+
+          super() # to init MonitorMixin
+        end
+
+        def start
+          synchronize do
+            return if @request_queue
+
+            @request_queue = EnumeratorQueue.new(self)
+            @thread_pool = Concurrent::FixedThreadPool.new(
+              threads,
+              fallback_policy: :caller_runs
+            )
+            start_streaming!
+          end
+
+          true
+        end
+
+        def stop
```

Comment:
I get this error if I #stop from a SIGINT trap:

```ruby
# ...
trap "INT" do
  puts "Exiting..."
  subscription.stop
  subscription.wait!
  exit
end
# ...
```

```
be ruby subscriber_test.rb 
Listening...
^CExiting...
/Users/remily/.rbenv/versions/2.3.1/lib/ruby/2.3.0/monitor.rb:187:in `lock': can't be called from trap context (ThreadError)
	from /Users/remily/.rbenv/versions/2.3.1/lib/ruby/2.3.0/monitor.rb:187:in `mon_enter'
	from /Users/remily/.rbenv/versions/2.3.1/lib/ruby/2.3.0/monitor.rb:212:in `mon_synchronize'
	from /Users/remi/Desktop/pubsub_review/gcloud-ruby/google-cloud-pubsub/lib/google/cloud/pubsub/subscriber.rb:69:in `stop'
	from subscriber_test.rb:19:in `block in <main>'
	from subscriber_test.rb:25:in `sleep'
	from subscriber_test.rb:25:in `block in <main>'
	from subscriber_test.rb:24:in `loop'
	from subscriber_test.rb:24:in `<main>'
```

I expect stopping from a trapped signal to be somewhat common (it's how I frequently use Pub/Sub, personally).